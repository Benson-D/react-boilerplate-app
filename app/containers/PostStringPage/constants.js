/*
 * Post String Constants
 */

export const ADD_STR_CHANGE = 'boilerplate/PostStringPage/ADD_STR_CHANGE';
export const ADD_STR_SUBMIT = 'boilerplate/PostStringPage/ADD_STR_SUBMIT';
export const ADD_STR_ERROR = 'boilerplate/PostStringPage/ADD_STR_ERROR';
