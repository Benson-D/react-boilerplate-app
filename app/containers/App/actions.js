/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import { LOAD_STR_DATA, LOAD_STR_SUCCESS, LOAD_STR_ERROR } from './constants';

/**
 * Load the string list, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_STR_DATA
 */
export function loadStrData() {
  return {
    type: LOAD_STR_DATA,
  };
}

/**
 * Dispatched when the strings are loaded by the request saga
 *
 * @param  {object} strings The string data
 * @return {object} An action object with a type of LOAD_STRINGS_SUCCESS passing the strings
 */
export function loadStrSuccess(strings) {
  return {
    type: LOAD_STR_SUCCESS,
    strings,
  };
}

/**
 * Dispatched when loading the string data fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_STR_ERROR passing the error
 */
export function loadStrError(error) {
  return {
    type: LOAD_STR_ERROR,
    error,
  };
}
