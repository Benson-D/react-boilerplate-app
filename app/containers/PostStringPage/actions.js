/*
 * PostStringPage Actions
 *
 */

import { ADD_STR_CHANGE, ADD_STR_SUBMIT, ADD_STR_ERROR } from './constants';

/**
 * Takes in the input field of the form
 *
 * @param  {string} string The new text of the input field
 *
 * @return {object} An action object with a type of ADD_STR_CHANGE
 */
export function changeString(string) {
  return {
    type: ADD_STR_CHANGE,
    string,
  };
}

/**
 * Updates the status of the form
 *
 * @param  {string} submit The final version of the input field
 *
 * @return {object} An action object with a type of ADD_STR_SUBMIT
 */
export function submitString(submit) {
  return {
    type: ADD_STR_SUBMIT,
    submit,
  };
}

/**
 * Dispatched when posting faces an error
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of ADD_STR_ERROR passing the error
 */
export function submitError(error) {
  return {
    type: ADD_STR_ERROR,
    error,
  };
}
