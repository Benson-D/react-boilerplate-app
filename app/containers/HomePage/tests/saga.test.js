/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_STR_DATA } from 'containers/App/constants';
import { loadStrSuccess, loadStrError } from 'containers/App/actions';

import serverStrData, { getStringData } from '../saga';

/* eslint-disable redux-saga/yield-effects */
describe('getStringData Saga', () => {
  let getStringGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getStringGenerator = getStringData();

    const callDescriptor = getStringGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadStrSuccess action if it requests the data successfully', () => {
    const response = ['test1', 'test2'];
    const putDescriptor = getStringGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(loadStrSuccess(response)));
  });

  it('should call the loadStrError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getStringGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(loadStrError(response)));
  });
});

describe('server Saga', () => {
  const serverApiDataSaga = serverStrData();

  it('should start task to watch for LOAD_STR_DATA action', () => {
    const takeLatestDescriptor = serverApiDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(LOAD_STR_DATA, getStringData),
    );
  });
});
