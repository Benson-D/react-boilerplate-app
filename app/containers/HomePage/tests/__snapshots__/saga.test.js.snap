// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStringData Saga should call the loadStrError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/strContainer",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getStringData Saga should dispatch the loadStrSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://localhost:3001/strContainer",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
